In [94]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

import sklearn

# Using Skicit-learn to split data into training and testing sets
from sklearn.model_selection import train_test_split


In [95]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
# evaluate random forest algorithm for classification
from numpy import mean
from numpy import std
from sklearn.datasets import make_classification
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import RepeatedStratifiedKFold, RepeatedKFold
from sklearn.ensemble import RandomForestClassifier

# Modelos mixtos. Admiten ser utilizados para regresión y clasificación

In [96]:
ENS_2017_C1 = pd.read_csv("../data/Bases_trabajo/ENS_2017_C1.csv", sep =',')

In [97]:
X = ENS_2017_C1[['CCAA', 'Sexo', 'Edad',
       'Actividad_economica_actual', 'Nacionalidad_española', 'Convivencia',
       'Estado_civil', 'Nivel_estudios', 'Vacunación_gripe',
       'Toma_tensiónArterial_profesional', 'Medición_colesterol',
       'Prueba_sangreHeces', 'Colonoscopia', 'Peso(Kg)',
       'Freq_ActividadFísica', 'Freq_Consumo_FrutaFresca',
       'Freq_Consumo_Carne', 'Freq_Consumo_Huevos', 'Freq_Consumo_Pescado',
       'Freq_Consumo_PastaArrozPatatas', 'Freq_Consumo_PanCereales',
       'Freq_Consumo_VerdurasEnsaladasHortalizas', 'Freq_Consumo_Legumbres',
       'Freq_Consumo_EmbutidosFiambres', 'Freq_Consumo_Lácteos',
       'Freq_Consumo_Dulces', 'Freq_Consumo_ComidaRápida',
       'Freq_Consumo_ZumoNatural', 'Freq_Diaria_CepilladoDientes',
       '¿Fuma actualmente', 'Freq_Consumo_Alcohol',
       'ApoyoAfectivoPersonal_AmigosFamiliares', 'ClaseSocial_BasadaOcupación',
       'Índice_MasaCorporal']]

In [98]:
y = ENS_2017_C1['Salud_percibida']

In [99]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=101)

## Utilización de los modelos para regresión

## Regression tree

In [100]:
dtr = DecisionTreeRegressor(max_depth=3)
dtr.fit(X_train, y_train)

DecisionTreeRegressor(max_depth=3)

In [101]:
dtr.score(X_train,y_train)

0.1748439487254927

In [102]:
predictions_train = dtr.predict(X_train)
errores_train = abs(predictions_train - y_train)
np.mean(errores_train)

0.4658883891235441

In [103]:
prop = 100 * (errores_train/y_train)
mape = np.mean(prop)
mape

32.87621581739204

In [104]:
import_dict = {
    'features': X_train.columns,
    'importance': dtr.feature_importances_
}

pd.DataFrame(import_dict).sort_values('importance', ascending=False)

features  importance
2                                       Edad    0.460764
14                      Freq_ActividadFísica    0.236383
3                 Actividad_economica_actual    0.155016
30                      Freq_Consumo_Alcohol    0.147836
0                                       CCAA    0.000000
25                       Freq_Consumo_Dulces    0.000000
21  Freq_Consumo_VerdurasEnsaladasHortalizas    0.000000
22                    Freq_Consumo_Legumbres    0.000000
23            Freq_Consumo_EmbutidosFiambres    0.000000
24                      Freq_Consumo_Lácteos    0.000000
27                  Freq_Consumo_ZumoNatural    0.000000
26                 Freq_Consumo_ComidaRápida    0.000000
19            Freq_Consumo_PastaArrozPatatas    0.000000
28              Freq_Diaria_CepilladoDientes    0.000000
29                         ¿Fuma actualmente    0.000000
31    ApoyoAfectivoPersonal_AmigosFamiliares    0.000000
32               ClaseSocial_BasadaOcupación    0.000000
20                  Freq_Consumo_PanCereales    0.000000
17                       Freq_Consumo_Huevos    0.000000
18                      Freq_Consumo_Pescado    0.000000
1                                       Sexo    0.000000
16                        Freq_Consumo_Carne    0.000000
15                  Freq_Consumo_FrutaFresca    0.000000
13                                  Peso(Kg)    0.000000
12                              Colonoscopia    0.000000
11                        Prueba_sangreHeces    0.000000
10                       Medición_colesterol    0.000000
9           Toma_tensiónArterial_profesional    0.000000
8                           Vacunación_gripe    0.000000
7                             Nivel_estudios    0.000000
6                               Estado_civil    0.000000
5                                Convivencia    0.000000
4                      Nacionalidad_española    0.000000
33                       Índice_MasaCorporal    0.000000

In [105]:
dtr = DecisionTreeRegressor(max_depth=3)
dtr.fit(X_test, y_test)

DecisionTreeRegressor(max_depth=3)

In [106]:
dtr.score(X_test,y_test)

0.18997371371748006

In [107]:
predictions_test = dtr.predict(X_test)
errores_test = abs(predictions_test - y_test)
np.mean(errores_test)

0.4649295005698889

In [108]:
prop = 100 * (errores_test/y_test)
mape = np.mean(prop)
mape

32.77518289833445

In [109]:
import_dict = {
    'features': X_test.columns,
    'importance': dtr.feature_importances_
}

pd.DataFrame(import_dict).sort_values('importance', ascending=False)

features  importance
2                                       Edad    0.561484
14                      Freq_ActividadFísica    0.204672
3                 Actividad_economica_actual    0.140300
30                      Freq_Consumo_Alcohol    0.058266
6                               Estado_civil    0.028871
13                                  Peso(Kg)    0.006406
0                                       CCAA    0.000000
25                       Freq_Consumo_Dulces    0.000000
22                    Freq_Consumo_Legumbres    0.000000
23            Freq_Consumo_EmbutidosFiambres    0.000000
24                      Freq_Consumo_Lácteos    0.000000
28              Freq_Diaria_CepilladoDientes    0.000000
26                 Freq_Consumo_ComidaRápida    0.000000
27                  Freq_Consumo_ZumoNatural    0.000000
20                  Freq_Consumo_PanCereales    0.000000
29                         ¿Fuma actualmente    0.000000
31    ApoyoAfectivoPersonal_AmigosFamiliares    0.000000
32               ClaseSocial_BasadaOcupación    0.000000
21  Freq_Consumo_VerdurasEnsaladasHortalizas    0.000000
17                       Freq_Consumo_Huevos    0.000000
19            Freq_Consumo_PastaArrozPatatas    0.000000
18                      Freq_Consumo_Pescado    0.000000
1                                       Sexo    0.000000
16                        Freq_Consumo_Carne    0.000000
15                  Freq_Consumo_FrutaFresca    0.000000
12                              Colonoscopia    0.000000
11                        Prueba_sangreHeces    0.000000
10                       Medición_colesterol    0.000000
9           Toma_tensiónArterial_profesional    0.000000
8                           Vacunación_gripe    0.000000
7                             Nivel_estudios    0.000000
5                                Convivencia    0.000000
4                      Nacionalidad_española    0.000000
33                       Índice_MasaCorporal    0.000000

## Random Forest Regressor

In [110]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error

model = RandomForestRegressor(n_estimators=500,
                               max_depth = 5,
                               random_state=42)

model.fit(X_train, y_train)

y_pred_reg = model.predict(X_train)
mean_absolute_error(y_train, y_pred_reg)

0.4390856682407461

In [111]:
model.score(X_train,y_train)

0.27038031029792975

In [112]:
model = RandomForestRegressor(n_estimators=500,
                               max_depth = 5,
                               random_state=42)

model.fit(X_test, y_test)

y_pred_reg = model.predict(X_test)
mean_absolute_error(y_test, y_pred_reg)

0.41731133892634237

In [113]:
model.score(X_test,y_test)

0.3762668233362213

## Utilización de los modelos para clasificación

## Classification tree

In [114]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn import metrics


tree_clf = DecisionTreeClassifier(random_state=42)
tree_clf.fit(X_train, y_train)
y_pred_tree = tree_clf.predict(X_train)
print("Accuracy:", metrics.accuracy_score(y_train, y_pred_tree))


Accuracy: 1.0


In [115]:
tree_clf = DecisionTreeClassifier(random_state=42)
tree_clf.fit(X_test, y_test)
y_pred_tree = tree_clf.predict(X_test)
print("Accuracy:", metrics.accuracy_score(y_test, y_pred_tree))



Accuracy: 1.0


## Random Forest Classifier

In [116]:
from sklearn.model_selection import KFold

# define the model
model = RandomForestClassifier()
# evaluate the model 
# THIS DOESN'T TRAIN THE MODEL
k_fold = RepeatedStratifiedKFold(n_splits=10, n_repeats=2, random_state=2)
n_scores = cross_val_score(model, X_train, y_train, scoring='accuracy', cv=k_fold, n_jobs=-1, error_score='raise')
# report performance
print(n_scores)
print('Accuracy: %.3f (%.3f)' % (mean(n_scores), std(n_scores)))

[0.70967742 0.67419355 0.68709677 0.67741935 0.71243942 0.67528271
 0.68982229 0.70113086 0.71728595 0.67689822 0.70483871 0.69516129
 0.7        0.69354839 0.68336026 0.68659128 0.67689822 0.69466882
 0.68982229 0.68497577]
Accuracy: 0.692 (0.013)


In [117]:
max(n_scores)

0.7172859450726979

In [118]:
min(n_scores)

0.6741935483870968

In [119]:
from sklearn.ensemble import RandomForestClassifier

rnd_clf = RandomForestClassifier(n_estimators = 500,
                                max_depth = 4,
                                random_state = 42)

rnd_clf.fit(X_train, y_train)

y_pred_rf = rnd_clf.predict(X_train)
print("Accuracy:", metrics.accuracy_score(y_train, y_pred_rf))

Accuracy: 0.6771068776235066


In [120]:
from sklearn.ensemble import RandomForestClassifier

rnd_clf = RandomForestClassifier(n_estimators = 500,
                                max_depth = 4,
                                random_state = 42)

rnd_clf.fit(X_test, y_test)

y_pred_rf = rnd_clf.predict(X_test)
print("Accuracy:", metrics.accuracy_score(y_test, y_pred_rf))

Accuracy: 0.684312459651388


### .....................................................

## Conclusion. Modelos mixtos (Regresión y clasificación)

### Regresión. Los dos modelos de regresión son los que obtienen peores resultados.

###     - Regression tree. Score_train = .1748, Score_test = .1899
###     - Random Forest Regressor. Score_train = .2703, Score_test = .3762

### Clasificación. Entre los dos modelos de clasificación, el de Classification tree es que mejor clasifica los registros.
###     - Classification tree. Accuracy_train = 1.0, Accuracy_test = 1.0
###     - Random Forest Classifier. Accuracy_train = .6771, Accuracy_test = .6843

### Las variables que tienen mayor peso al realizar la clasificación son, por orden de importancia: Edad, Frecuencia de Actividad física, Actividad económica actual, Frecuencia de consumo de alcohol. En la muestra de test se incluyen después de las indicadas: Estado civil y Frecuencia de consumo de verduras, ensaladas y hortalizas